File: answers/leetcode/Surrounded Regions.cpp
class Solution {
public:
    void solve(vector<vector<char>> &board) {
		for (size_t i = 0; i < board.size(); ++ i) {
			for (size_t j = 0; j < board[i].size(); ++ j) {
				if ('x' == board[i][j] || 'X' == board[i][j] || '1' == board[i][j]) {
					continue;
				}
				if (0 == i || board.size() - 1 == i || 0 == j || board[i].size() - 1 == j) {
					board[i][j] = '1';
				}
				else if ('1' == board[i - 1][j] || '1' == board[i + 1][j] || '1' == board[i][j - 1] || '1' == board[i][j + 1]) {
					board[i][j] = '1';
				}
			}
		}
		for (size_t i = 0; i < board.size(); ++ i) {
			for (size_t j = 0; j < board[i].size(); ++ j) {
				size_t x = board.size() - i - 1;
				size_t y = board[i].size() - j - 1;
				if ('x' == board[x][y] || 'X' == board[x][y] || '1' == board[x][y]) {
					continue;
				}
				if (0 == x || board.size() - 1 == x || 0 == y || board[x].size() - 1 == y) {
					board[x][y] = '1';
				}
				else if ('1' == board[x - 1][y] || '1' == board[x + 1][y] || '1' == board[x][y - 1] || '1' == board[x][y + 1]) {
					board[x][y] = '1';
				}
			}
		}
		for (size_t i = 0; i < board.size(); ++ i) {
			for (size_t j = 0; j < board[i].size(); ++ j) {
				if ('x' == board[i][j] || 'X' == board[i][j] || '1' == board[i][j]) {
					continue;
				}
				if (0 == i || board.size() - 1 == i || 0 == j || board[i].size() - 1 == j) {
					dfs(i, j, board);
				}
				else if ('1' == board[i - 1][j] || '1' == board[i + 1][j] || '1' == board[i][j - 1] || '1' == board[i][j + 1]) {
					dfs(i, j, board);
				}
			}
		}
		for (size_t i = 0; i < board.size(); ++ i) {
			for (size_t j = 0; j < board[i].size(); ++ j) {
				if ('O' == board[i][j]) {
					board[i][j] = 'X';
				}
				else if ('1' == board[i][j]) {
					board[i][j] = 'O';
				}
			}
		}
    }
	void dfs(size_t x, size_t y, vector<vector<char> > &board) {
		if (0 == x || board.size() - 1 == x || 0 == y || board[x].size() - 1 == y) {
			return;
		}
		if ('x' == board[x][y] || 'X' == board[x][y] || '1' == board[x][y]) {
			return;
		}
		board[x][y] = '1';
		dfs(x - 1, y, board);
		dfs(x + 1, y, board);
		dfs(x, y - 1, board);
		dfs(x, y + 1, board);
	}
};

File: answers/leetcode/Surrounded Regions/Surrounded Regions.java
class Solution {
    public void solve(char[][] board) {
        if (board == null || board.length == 0 || board[0].length == 0) {
            return;
        }
        for (int i = 0; i < board.length; i++) {
            this.dfs(board, i, 0);
            this.dfs(board, i, board[i].length - 1);
        }
        for (int i = 0; i < board[0].length; i++) {
            this.dfs(board, 0, i);
            this.dfs(board, board.length - 1, i);
        }
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                } else if (board[i][j] == 'A') {
                    board[i][j] = 'O';
                }
            }
        }
    }

    private void dfs(char[][] board, int x, int y) {
        if (x < 0 || x >= board.length || y < 0 || y >= board[x].length || board[x][y] != 'O') {
            return;
        }
        board[x][y] = 'A';
        this.dfs(board, x - 1, y);
        this.dfs(board, x + 1, y);
        this.dfs(board, x, y - 1);
        this.dfs(board, x, y + 1);
    }
}